src/main/java/gregtech/api/capability/impl/CombinedCapabilityProvider.java

```diff
@@ -42,4 +44,33 @@ public <T> T getCapability(@NotNull Capability<T> capability, @Nullable EnumFaci
         }
         return null;
     }
+
+    @Override
+    public NBTTagCompound serializeNBT() {
+        NBTTagCompound tag = new NBTTagCompound();
+
+        int index = 0;
+        for (ICapabilityProvider provider : providers) {
+            if (provider instanceof INBTSerializable<?>serializable) {
+                tag.setTag(String.valueOf(index++), serializable.serializeNBT());
+            }
+        }
```

Comment:
i'd probably use an indexed `for` loop since that's basically what you're doing anyways